src/features/broker-protection/actions/extract.js

```diff
@@ -27,18 +27,44 @@ import { matchAddressFromAddressListCityState } from '../comparisons/address.js'
  * @param {Record<string, any>} userData
  * @return {import('../types.js').ActionResponse}
  */
+export function extract (action, userData) {
+    const extractResult = extractProfiles(action, userData)
+
+    if ('error' in extractResult) {
+        return new ErrorResponse({ actionID: action.id, message: extractResult.error })
+    }
+
+    const filtered = extractResult.results
+        .filter(x => x.result === true)
+        .map(x => aggregateFields(x.scrapedData))
+
+    return new SuccessResponse({ actionID: action.id, actionType: action.actionType, response: filtered })
+}
+
+/**
+ * @param {Action} action
+ * @param {Record<string, any>} userData
+ * @return {{error: string}|{results: {scrapedData: Record<string, any>, result: boolean, score: number, element: HTMLElement}[]}}
+ */
 export function extractProfiles (action, userData) {
     const profilesElementList = getElements(document, action.selector) ?? []
 
-    const matchedProfiles = profilesElementList
-        // first, convert each profile element list into a profile
-        .map((element) => createProfile(element, action.profile))
-        // only include profiles that match the user data
-        .filter((scrapedData) => scrapedDataMatchesUserData(userData, scrapedData))
-        // aggregate some fields
-        .map((scrapedData) => aggregateFields(scrapedData))
+    if (profilesElementList.length === 0) {
+        return { error: 'no root elements found for ' + action.selector }
+    }
 
-    return new SuccessResponse({ actionID: action.id, actionType: action.actionType, response: matchedProfiles })
+    return {
+        results: profilesElementList.map((element) => {
+            const scrapedData = createProfile(element, action.profile)
+            const { result, score } = scrapedDataMatchesUserData(userData, scrapedData)
+            return {
+                scrapedData,
+                result,
+                score,
+                element
+            }
+        })
+    }
```

Comment:
this change is all about returning all elements, to allow different consumers to choose which element to use. 

For example, a scan might be really loose and just select multiple matches, but other places like a search situation you might want to consider the match's score

Reply:
This is well thought-through, nicely done. I love the flexibility of letting the consumer decide how to use the results.